test/cpp/end2end/xds/xds_cluster_end2end_test.cc

```diff
@@ -304,6 +311,58 @@ TEST_P(CdsTest, ClusterChangeAfterAdsCallFails) {
   WaitForBackend(DEBUG_LOCATION, 1);
 }
 
+MATCHER(VerifyCsmServiceLabels, "") {
+  using OptionalLabelComponent =
+      grpc_core::ClientCallTracer::CallAttemptTracer::OptionalLabelComponent;
+  auto iter = arg.find(OptionalLabelComponent::kXdsServiceLabels);
+  if (iter != arg.end()) {
+    const auto& xds_service_labels_map = iter->second;
+    if (xds_service_labels_map.find("service_name") !=
+            xds_service_labels_map.end() &&
+        xds_service_labels_map.find("service_namespace") !=
+            xds_service_labels_map.end()) {
+      return xds_service_labels_map.at("service_name") == "myservice" &&
+             xds_service_labels_map.at("service_namespace") == "mynamespace";
+    }
+  }
+  return false;
+}
+
+TEST_P(CdsTest, VerifyCsmServiceLabelsParsing) {
+  CreateAndStartBackends(1);
+  // Populates EDS resources.
+  EdsResourceArgs args({{"locality0", CreateEndpointsForBackends()}});
+  balancer_->ads_service()->SetEdsResource(
+      BuildEdsResource(args, kDefaultClusterName));
+
+  // Populates service labels to CDS resources.
+  auto cluster = default_cluster_;
+  google::protobuf::Struct labels;
```

Comment:
Same comment as in xds_cluster_resource_type_test.cc about writing this more idomatically and with less copying.